### PR TITLE
#5588: gh post commit job runs unevenly on ci runners

### LINKED ISSUE #5588: GH post-commit job runs unevenly on ci runners
Workflow jobs can run better if test suites are redistributed and balanced to handle workloads better.
Should reduce runtime by moving workloads to run faster on TT CI Runners

### PR DESCRIPTION
**Pull request by ttmchiou**
Rebalanced workflows to run in a more readable manner and redistributed fast-dispatch unit tests so that workflow runs faster.
An additional 3 more runners were be required to cut down on post-commit run time for fast-dispatch
Max time reduced down to ~1 hour

Additional Intent:
- Split out the job from 1 step into multiple steps. This will let devs have a more intuitive view of which file generated the crash. File logs will probably be still long but there are technically more points to check the log.
- Devs should still be able to run normally without being affected from these changes except for maybe needing to be aware of how to add new pytests for eager tests? 
- Took advantage of the fact that this will be always fast dispatch so tests are individually called into of using scripts since it was a little messy to split here.

Note:
These changes will indirectly trigger more issues in how we execute workflows as it bypasses calling run_scripts directly. though functionality is still preserved if developers wish to call it on cli on their local machines.
Open to thoughts/opinions on this.

Workflow Example (only fast dispatch)
https://github.com/tenstorrent-metal/tt-metal/actions/runs/8018803993

Full Post-commit run:
https://github.com/tenstorrent-metal/tt-metal/actions/runs/8018807314

### CODE REVIEW

**Comment by ttmchiou** [comment]
draft status pending until post-commit runs pass successfully

**Review by TT-billteng — APPROVED** [review]

**Review by tt-rkim — COMMENTED** [review]
- Make us codeowners of `tests/scripts/run_pre_post_commit_regressions_fast_dispatch.sh` in `/CODEOWNERS` 
- Add docs in `CONTRIBUTING.md`
  - Change "Running tests in tt-metal" to "Tests in tt-metal"
  - Add new section about adding post commit tests

we're gucci

File: .github/workflows/fast-dispatch-build-and-unit-tests.yaml [hunk 1]
```diff
@@ -34,8 +35,76 @@ jobs:
...
+          env python tests/scripts/run_tt_metal.py --dispatch-mode fast
+          env python tests/scripts/run_tt_eager.py --dispatch-mode fast
+      - name: Run pre/post C++ regression tests
+        timeout-minutes: 120
+        run: |
+          source build/python_env/bin/activate
+          cd $TT_METAL_HOME
+          export PYTHONPATH=$TT_METAL_HOME
+          ./build/test/tt_metal/unit_tests_fast_dispatch
+      - name: Build Docs
```
[the remark this answers is not in the document]
Comment (TT-billteng) [reply .github/workflows/fast-dispatch-build-and-unit-tests.yaml:104]: is this basically tested [here](https://github.com/tenstorrent-metal/tt-metal/actions/workflows/docs-latest-public.yaml) ?

**Comment by ttmchiou** [comment]
@TT-billteng 
@tt-rkim 
Made changes according to comments,
Left the build docs step in as we do want to make sure we keep building docs during post-commit runs.

File: .github/workflows/fast-dispatch-build-and-unit-tests.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (tt-rkim) [reply .github/workflows/fast-dispatch-build-and-unit-tests.yaml:103]: @ttmchiou @TT-billteng 
Yes, it is tested in that github runenr workflow, but we didn't add it to the all post commits workflow.

I think there's some sussy stuff about it running on branches, but shouldn't be too hard to fix...

If we want to merge this soon, we could make a separate issue to fix the docs pipeline on branch, delete this step, and then add the docs pipeline to all post commit workflow.

**Review by tt-rkim — APPROVED** [review]

**Comment by ttmchiou** [comment]
Merging.
Completed run here
https://github.com/tenstorrent-metal/tt-metal/actions/runs/8072598531/job/22057660328

**Pull request merged**